File: socket/singleServer.js
var { SocketType, NotificationType, RequestMethod } = require('./constans');
const socketIo = require('socket.io');
var app = require('express')();
//通过http模块Server方法获取关联Express的服务对象
var server = require('http').createServer(app);
//导入socket.io模块获取io对象并关联http服务对象
var io = socketIo(server, {
	cors: {
		origin: 'http://localhost:8889', // 允许的来源地址
		methods: ['GET', 'POST'], // 允许的方法
		credentials: true, // 是否允许发送cookies等凭证
	},
});

var url = 'http://localhost:9001';

var port = 9001;

var users = {};

var messages = [];

var msgId = 1;

io.on('connection', function (socket) {
	const user = socket.handshake.query;
	console.log('start connection: ', user);

	const userOpt = {
		id: socket.id,
		username: user.name,
	};
	users[userOpt.id] = userOpt;

	sendSelfNotification(NotificationType.roomReady, { messages, users: Object.values(users) });
	sendOtherUsersNotification(NotificationType.newUser, userOpt);

	//服务端监听客户端发送的新消息
	socket.on('request', function ({ method, data }, cb) {
		const socketQuery = socket.handshake.query;
		if (method === RequestMethod.leave) {
			delete users[socket.id];
			sendOtherUsersNotification(NotificationType.userClosed, socket.id);
			cb(null, { code: 200, message: 'User left' }); // 确保调用回调函数
		} else if (method === RequestMethod.sendMsg) {
			const opt = { username: socketQuery.name, message: data, id: msgId };
			messages.push(opt);
			msgId++;
			sendOtherUsersNotification(NotificationType.newMessage, opt);
			cb(null, { code: 200, message: 'Message sent' }); // 确保调用回调函数
		} else {
			cb(new Error('Unknown method'), null); // 处理未知方法的情况
		}
	});

	socket.on('disconnect', function () {
		const socketQuery = socket.handshake.query;
		console.log('disconnect: ', socketQuery);
		delete users[socket.id];
		sendOtherUsersNotification(NotificationType.userClosed, socket.id);
	});

	//发送给其他人的消息
	function sendOtherUsersNotification(method, data) {
		console.log('sendOtherUsersNotification method data', method, data);
		socket.broadcast.emit('notification', { method, data });
	}

	//发送给自己的信息
	function sendSelfNotification(method, data) {
		console.log('sendSelfNotification method data', method, data);
		socket.emit('notification', { method, data });
	}
});

server.listen(port, function () {
	console.log('App listening on port 9001!');
});

File: socket/singleServer.py
from typing import Dict, List
from flask import Flask, request
from flask_socketio import SocketIO, emit

app = Flask(__name__)
# socketio = SocketIO(app, cors_allowed_origins="http://localhost:8999")
socketio = SocketIO(app, cors_allowed_origins="*")

# 服务端信息枚举
class ServerMessageEnum():
    notification = 'notification'
    request = 'request'

# 通知客户端信息枚举
class NotificationEnum():
    newUser = 'newUser'
    roomReady = 'roomReady'
    userClosed = 'userClosed'
    newMessage = 'newMessage'

# 请求客户端信息枚举
class RequestEnum():
    leave = 'leave'
    sendMsg = 'sendMsg'

# 用户信息类
class User:
    def __init__(self, id: str, username: str):
        self.id = id
        self.username = username

    def to_dict(self):
        return {
            'id': self.id,
            'username': self.username
        }

# 消息信息类
class Message:
    def __init__(self, username: str, message: str, id: int):
        self.username = username
        self.message = message
        self.id = id

    def to_dict(self):
        return {
            'username': self.username,
            'message': self.message,
            'id': self.id
        }

# Socket.io 响应类
class SocketIoRes:
    def __init__(self, method: str, data):
        self.method = method
        self.data = data

    def to_dict(self):
        # 确保 data 是可序列化的
        if hasattr(self.data, 'to_dict'):
            data = self.data.to_dict()
        elif isinstance(self.data, (list, dict, str, int, float, bool, type(None))):
            data = self.data
        elif isinstance(self.data, (tuple, set)):
            data = list(self.data)
        else:
            data = str(self.data)
        return {
            'method': self.method,
            'data': data
        }

users: Dict[str, User] = {}
messages: List[Message] = []
msg_id = 1

@socketio.on('connect')
def handle_connect(auth=None):  # 添加 auth 参数
    user_name = request.args.get('name')
    if not user_name:
        socketio.disconnect(request.sid)
        print("Connection denied: No username provided")
        return
    user = User(request.sid, user_name)
    users[user.id] = user
    newUserRes = SocketIoRes(NotificationEnum.newUser, user.to_dict())
    emit(ServerMessageEnum.notification, newUserRes.to_dict(), broadcast=True)
    users_dict = [user.to_dict() for user in users.values()]
    messages_dict = [message.to_dict() for message in messages]
    roomReadyRes = SocketIoRes(NotificationEnum.roomReady, {'messages': messages_dict, 'users': users_dict})
    emit(ServerMessageEnum.notification, roomReadyRes.to_dict(), broadcast=True)
    print(f"start connection: {user}")

@socketio.on(ServerMessageEnum.request)
def handle_request(data):
    global msg_id
    res = SocketIoRes(data['method'], data['data'])
    if res.method == RequestEnum.leave:
        user_id = request.sid
        if user_id in users:
            user_opt = users[user_id]
            del users[user_id]
            userClosedRes = SocketIoRes(NotificationEnum.userClosed, user_opt.to_dict())
            emit(ServerMessageEnum.notification, userClosedRes.to_dict(), broadcast=True)
            print(f"disconnect: {user_opt}")
        socketio.close_room(user_id)
    elif res.method == RequestEnum.sendMsg:
        user_id = request.sid
        if user_id not in users:
            print(f"User not found: {user_id}")
            return
        user_name = users[user_id].username
        opt = Message(user_name, res.data, msg_id)
        messages.append(opt)
        msg_id += 1
        newMessageRes = SocketIoRes(NotificationEnum.newMessage, opt.to_dict())
        emit(ServerMessageEnum.notification, newMessageRes.to_dict(), broadcast=True)
        print(f"new message: {opt}")
    else:
        print("Unknown method")

@socketio.on('disconnect')
def handle_disconnect():
    user_id = request.sid
    if user_id in users:
        user_opt = users[user_id]
        del users[user_id]
        userClosedRes = SocketIoRes(NotificationEnum.userClosed, user_opt.to_dict())
        emit(ServerMessageEnum.notification, userClosedRes.to_dict(), broadcast=True)
        print(f"disconnect: {user_opt}")

if __name__ == '__main__':
    socketio.run(app, host='0.0.0.0', port=9001)
